S9-TEST1

lyrapanged opened the pull request:
L1

lyrapanged committed:
refactor(User,UserController,Film,FilmController)

lyrapanged committed:
S9-TEST1

lyrapanged commented:
L1

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private void validationUser(User user) {
+        if (!(user.getEmail().contains("@")) || user.getEmail().isBlank()) {
```
Для начала следует проверить, что строка не равна null, иначе есть шанс возникновения NPE, если она вообще не будет передана)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,68 @@
+            log.error("Bad email");
+            throw new ValidationException("Email cannot be empty and contain the symbol \"@\".");
+        }
+        if (user.getLogin().isEmpty() || user.getLogin().contains(" ")) {
```
Для начала следует проверить, что логин не равен null, иначе есть шанс возникновения NPE, если он вообще не будет передан) Так же хорошо было бы использовать не isEmpty, а isBlank, так как первый засчитает за не пустую строку набор пробелов и валидация будет пройдена успешно)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,68 @@
+            log.error("Bad login");
+            throw new ValidationException("Login cannot be empty and contain spaces.");
+        }
+        if (user.getName() == null) {
```
Так же следует проверить имя с помощью isBlank, так как в данном случае так же следует установить имя == логину

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,68 @@
+            log.info("Name has been set to Login");
+            user.setName(user.getLogin());
+        }
+        if (!user.getBirthday().isBefore(LocalDate.now())) {
```
Для начала следует проверить, что полученная дата не null, иначе есть шанс возникновения NPE

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private void validationFilm(Film film) {
+        if (film.getName().isBlank()) {
```
Для начала следует проверить, что имя не null, иначе есть шанс возникновения NPE, если оно вообще не будет передано)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,71 @@
+            log.error("Bad description.");
+            throw new ValidationException("Max description length is 200 symbols.");
+        }
+        if (film.getReleaseDate().isBefore(LOWER_DATE)) {
```
Для начала следует проверить, что дата не null, иначе есть шанс возникновения NPE, если она вообще не будет передана)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,71 @@
+            log.error("Bad name.");
+            throw new ValidationException("Name cannot be empty.");
+        }
+        if (film.getDescription().length() > DESCRIPTION_LENGTH) {
```
Для начала следует проверить, что описание не null, иначе есть шанс возникновения NPE, если оно вообще не будет передано)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,16 @@
+
+@Data
+@Builder
+public class Film {
```
(Предложение)Можно было бы избавиться от большинства проверок валидации в коде и реализовать их с помощью аннотаций) Например, над именем можно было бы использовать аннотацию @NotBlank, которая пресечет передачу null либо же название только из пробелов) 
Над описанием можно было бы использовать @Size, а для даты релиза использовать @NotNull, чтобы запретить не передавать ее) Для продолжительности @Positive или @Min(1), чтобы запретить передавать отрицательное значение) Читобы эти аннотации стали учитываться, следует добавить перед параметром Film film, в методах создания и обновления, аннотацию @Valid. Аналогично и с пользователем)

TyutterinYakov reviewed (commented):
Добрый день! Бинарные файлы и ненужные файлы конфигурации по типу mvnw.cmd, mvnw и тд можно было бы поместить в .gitignore. В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы - пиши в слак, буду рад помочь)

lyrapanged committed:
fix(User,UserController,Film,FilmController) adding NPE check/S9-TEST1
refactor(User,UserController,Film,FilmController) adding @Validation/S9-TEST1
docs(gitignore) adding mvnw.*

lyrapanged committed:
fix(User,UserController,Film,FilmController) adding NPE check/S9-TEST1
refactor(User,UserController,Film,FilmController) adding @Validation/S9-TEST1
docs(gitignore) adding mvnw.*
fix(User,film)

lyrapanged committed:
fix(User,UserController,Film,FilmController) adding NPE check/S9-TEST1
refactor(User,UserController,Film,FilmController) adding @Validation/S9-TEST1
docs(gitignore) adding mvnw.*
fix(User,film)
Adding Description_length(FilmController)

lyrapanged commented:
(FilmsController)По неизвестной мне причине, при добавлении к полю String description, аннотации @Size(max = 200) ,тесты успешно выполняются в Postman, но GitHub Actions пропускает объект у которого description > 200, пришлось оставить константу DESCRIPTION_LENGH и провести валидацию описания в ручную.

TyutterinYakov commented on src/test/java/ru/yandex/practicum/filmorate/controllersTest/FilmControllerTest.java:
```diff
@@ -0,0 +1,57 @@
+
+class FilmControllerTest {
+
+    private final FilmController filmController = new FilmController();
```
Для внедрения контроллера следует использовать @Autowired
Самостоятельно объект этого класса мы не создаем, этим занимается Spring)

TyutterinYakov commented on src/test/java/ru/yandex/practicum/filmorate/controllersTest/UserControllerTest.java:
```diff
@@ -0,0 +1,61 @@
+            .birthday(LocalDate.of(2020, 2, 22))
+            .build();
+
+    private UserController userController = new UserController();
```
Для внедрения контроллера следует использовать @Autowired
Самостоятельно объект этого класса мы не создаем, этим занимается Spring)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,23 @@
+@Builder
+public class User {
+    private Integer id;
+    @NonNull
```
В данном случае следует использовать @NotNull, но нам и она не требуется, так как она входит в @NotBlank)
https://www.programmersought.com/article/35893798426/
Так же можно было бы использовать аннотацию @Email, чтобы избавиться от проверки в коде)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,23 @@
+    @NonNull
+    @NotBlank
+    private String email;
+    @NonNull
```
В данном случае следует использовать @NotNull, но нам и она не требуется, так как она входит в @NotBlank)
https://www.programmersought.com/article/35893798426/

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,23 @@
+    @NotBlank
+    private String login;
+    private String name;
+    @NonNull
```
В данном случае следует использовать @NotNull
https://www.programmersought.com/article/35893798426/
Так же хорошо было бы использовать аннотацию @PastOrPresent, которая допускает только дату из прошлого и текущую

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,27 @@
+    @NotBlank
+    @Size(max = 200)
+    private String description;
+    @NonNull
```
В данном случае следует использовать @NotNull
https://www.programmersought.com/article/35893798426/

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private void validationUser(@Valid User user) {
+        if (!(user.getEmail().contains("@"))) {
```
Проверка почты не требуется, так как она реализована с помощью аннотации @Email)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,72 @@
+        return new ArrayList<>(films.values());
+    }
+
+    private void validationFilm(@Valid @RequestBody Film film) {
```
Аннотации @Valid @RequestBody используются в методах-обработчиках, здесь они не требуются)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,69 @@
+        return new ArrayList<>(users.values());
+    }
+
+    private void validationUser(@Valid User user) {
```
Аннотация @Valid используется в методах-обработчиках, здесь она не требуется)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,72 @@
+            log.error("Bad release date");
+            throw new ValidationException("Release date - no earlier than December 28, 1895.");
+        }
+        if (film.getDuration() <= 0) {
```
Проверка реализована с помощью аннотаций

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,69 @@
+            log.info("Name has been set to Login");
+            user.setName(user.getLogin());
+        }
+        if (!user.getBirthday().isBefore(LocalDate.now())) {
```
Дата теперь проверяется с помощью аннотаций)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,72 @@
+            log.error("Bad name.");
+            throw new ValidationException("Name cannot be empty.");
+        }
+        if (film.getDescription().length() > DESCRIPTION_LENGTH) {
```
Проверка описания не требуется, так как она реализована с помощью аннотаций

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private void validationFilm(@Valid @RequestBody Film film) {
+        if (film.getName().isBlank()) {
```
Проверка названия фильма так же не требуется)

TyutterinYakov reviewed (commented):
Добрый вечер! Думаю, что проблема у тебя в том, что ты пытаешься валидацию с помощью аннотаций протестировать с помощью Junit, что не совсем верно. В будущем ты познакомишься с другими видами тестирования, а пока что все тесты, для которых валидация реализована с помощью аннотаций - следует убрать. Хорошего вечера и отличного настроения :)

johnskyi commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -0,0 +1,64 @@
+@RestController
+@RequestMapping("/films")
+@Slf4j
+@Component
```
@Component эта аннотация лишняя здесь. @RestController ее уже включает

johnskyi commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -0,0 +1,62 @@
+@RestController
+@RequestMapping("/users")
+@Slf4j
+@Component
```
аналогично. лишняя аннотация

johnskyi commented on src/test/java/ru/yandex/practicum/filmorate/controllersTest/FilmControllerTest.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.time.LocalDate;
+
+@Component
```
Эта аннотация тут тоже не нужна) не лепите ее пожалуйста везде

lyrapanged committed:
refactor(User,UserController,Film,FilmController) remove @Component
